apps/oracledb.go

```diff
@@ -859,12 +861,25 @@ func (lr LoggingProcessorOracleDBAlert) Components(tag string, uid string) []flu
 type LoggingReceiverOracleDBAlert struct {
 	LoggingProcessorOracleDBAlert           `yaml:",inline"`
 	confgenerator.LoggingReceiverFilesMixin `yaml:",inline" validate:"structonly"`
+	OracleHome                              string `yaml:"oracle_home,omitempty"`
+}
+
+func (lr LoggingReceiverOracleDBAlert) CustomValidate(id string) error {
+	if len(lr.OracleHome) > 0 && len(lr.IncludePaths) > 0 {
+		return fmt.Errorf("\"oracle_home\" and \"include_paths\" cannot both be specified for %s receiver \"%s\"", lr.Type(), id)
+	}
+
+	return nil
 }
 
 func (lr LoggingReceiverOracleDBAlert) Components(tag string) []fluentbit.Component {
-	if len(lr.IncludePaths) == 0 {
+	if len(lr.OracleHome) > 0 {
+		lr.IncludePaths = []string{
+			fmt.Sprintf("%s/diag/rdbms/*/*/trace/alert_*.log", lr.OracleHome),
+		}
+	} else if len(lr.IncludePaths) == 0 {
 		lr.IncludePaths = []string{
-			"$ORACLE_HOME/diag/rdbms/*/*/trace/alert_*.log",
+			"${ORACLE_HOME}/diag/rdbms/*/*/trace/alert_*.log",
```

Comment:
What I had in mind was to require exactly one of `oracle_home` or `include_paths` since we can't rely on `$ORACLE_HOME` in any circumstance -- if we tell the user they need `$ORACLE_HOME` anyway in the default case then it defeats the purpose of having `oracle_home`.

I tested something like this using built-in validators and it seemed to work, and it would eliminate the need for any custom validator code:

```go
type LoggingReceiverOracleDBAlert struct {
	LoggingProcessorOracleDBAlert           `yaml:",inline"`
	confgenerator.LoggingReceiverFilesMixin `yaml:",inline" validate:"structonly"`
	OracleHome                              string   `yaml:"oracle_home,omitempty" validate:"required_without=IncludePaths,excluded_with=IncludePaths"`
	IncludePaths                            []string `yaml:"include_paths,omitempty" validate:"required_without=OracleHome,excluded_with=OracleHome"`
}
```

Since `IncludePaths` is shadowing the mixin field, we have to copy it to the mixin before generating its components:

```go
func (lr LoggingReceiverOracleDBAlert) Components(tag string) []fluentbit.Component {
	if len(lr.OracleHome) > 0 {
		lr.IncludePaths = []string{
			fmt.Sprintf("%s/diag/rdbms/*/*/trace/alert_*.log", lr.OracleHome),
		}
	}
	lr.LoggingReceiverFilesMixin.IncludePaths = lr.IncludePaths
	c := lr.LoggingReceiverFilesMixin.Components(tag)
	c = append(c, lr.LoggingProcessorOracleDBAlert.Components(tag, lr.Type())...)
	return c
}
```

Not sure if there's a cleaner way to do the above, but I would definitely prefer using the built-in validators if possible.

Reply:
That's great, I couldn't quite figure out a way to get it hooked up right, this looks like it works perfectly.